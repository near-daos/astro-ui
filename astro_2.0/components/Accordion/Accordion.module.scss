@import 'styles/typography';

$transition-time: 0.3s;

.root {
  border-radius: 0 8px 8px;
  box-shadow: 3px 2px 24px rgba(0, 0, 0, 0.1);
  padding: 0 24px;
}

.header {
  @extend %title2;
  align-items: center;
  cursor: pointer;
  display: flex;
  height: 60px;
  justify-content: space-between;
}

.icon {
  transition: $transition-time;
}

.open .icon {
  transform: rotate(180deg);
}

.content {
  margin-bottom: 22px;
}

.contentContainer {
  overflow: auto;
  transition: $transition-time;
}
