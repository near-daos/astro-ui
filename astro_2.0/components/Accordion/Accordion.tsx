import cn from 'classnames';
import React, { FC, ReactNode, useEffect, useRef, useState } from 'react';

import { IconButton } from 'components/button/IconButton';

import styles from './Accordion.module.scss';

interface AccordionProps {
  title: ReactNode;
  isOpen?: boolean;
  className?: string;
}

export const Accordion: FC<AccordionProps> = props => {
  const { title, children, isOpen = false, className } = props;

  const [open, setOpen] = useState(isOpen);

  const contentContainer = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const contentContainerEl = contentContainer.current;

    if (contentContainerEl) {
      contentContainerEl.style.height = open
        ? `${contentContainerEl.scrollHeight}px`
        : '0';
    }
  }, [open]);

  function toggleOpenState() {
    setOpen(!open);
  }

  const rootClassName = cn(styles.root, className, {
    [styles.open]: open,
  });

  return (
    <div className={rootClassName}>
      <div
        tabIndex={0}
        role="button"
        className={styles.header}
        onClick={toggleOpenState}
        onKeyPress={toggleOpenState}
      >
        {title}
        <IconButton
          size="medium"
          icon="buttonArrowDown"
          className={styles.icon}
        />
      </div>
      <div ref={contentContainer} className={styles.contentContainer}>
        <div className={styles.content}>{children}</div>
      </div>
    </div>
  );
};
