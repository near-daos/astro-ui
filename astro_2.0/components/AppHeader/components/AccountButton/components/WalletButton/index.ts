export * from 'astro_2.0/components/AppHeader/components/AccountButton/components/WalletButton/WalletButton';
