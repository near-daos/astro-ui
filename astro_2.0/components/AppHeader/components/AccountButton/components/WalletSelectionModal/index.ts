export * from 'astro_2.0/components/AppHeader/components/AccountButton/components/WalletSelectionModal/WalletSelectionModal';
