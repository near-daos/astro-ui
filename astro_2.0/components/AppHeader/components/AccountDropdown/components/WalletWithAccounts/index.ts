export * from 'astro_2.0/components/AppHeader/components/AccountDropdown/components/WalletWithAccounts/WalletWithAccounts';
