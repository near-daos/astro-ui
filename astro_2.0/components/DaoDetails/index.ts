export * from './DaoDetailsMinimized';
export * from './DaoDetailsGrid';
