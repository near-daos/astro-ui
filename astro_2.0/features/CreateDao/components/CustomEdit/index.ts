import CustomEdit from './CustomEdit';

export default CustomEdit;
