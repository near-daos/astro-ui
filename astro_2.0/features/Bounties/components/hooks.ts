import { useCallback, useRef, useState } from 'react';
import { useRouter } from 'next/router';
import { useAsyncFn, useLocalStorage, useLocation } from 'react-use';
import axios, { CancelTokenSource } from 'axios';

import { SputnikHttpService } from 'services/sputnik';
import { useModal } from 'components/modal';
import { ConfirmActionModal } from 'astro_2.0/components/ConfirmActionModal';

import { Bounty, BountyContext, BountyProposal } from 'types/bounties';
import { VoteAction } from 'types/proposal';

import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';

import { useWalletContext } from 'context/WalletContext';
import { PaginationResponse } from 'types/api';
import { VOTE_ACTION_SOURCE_PAGE } from 'constants/votingConstants';

export function useBountyControls(
  daoId: string,
  bounty?: Bounty
): {
  handleClaim: () => void;
  handleUnclaim: () => void;
  loading: boolean;
} {
  const router = useRouter();
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();
  const [loading, setLoading] = useState(false);

  const [showModal] = useModal(ConfirmActionModal);

  const onSuccessHandler = useCallback(async () => {
    router.reload();
  }, [router]);

  const handleClaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Claim',
    });

    if (res?.length && bounty) {
      setLoading(true);

      const dao = await SputnikHttpService.getDaoById(daoId);

      if (!dao) {
        return;
      }

      try {
        setVoteActionSource(pathname);

        await nearService?.claimBounty(daoId, {
          bountyId: bounty?.bountyId,
          deadline: bounty?.maxDeadline,
          bountyBond: dao.policy.bountyBond,
          gas: res[0],
          tokenId: bounty?.token,
        });

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    daoId,
    setVoteActionSource,
    pathname,
    nearService,
    onSuccessHandler,
  ]);

  const handleUnclaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Unclaim',
    });

    if (res?.length && bounty) {
      try {
        setLoading(true);
        setVoteActionSource(pathname);
        await nearService?.unclaimBounty(daoId, bounty?.bountyId);

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    setVoteActionSource,
    pathname,
    nearService,
    daoId,
    onSuccessHandler,
  ]);

  return {
    handleClaim,
    handleUnclaim,
    loading,
  };
}

export function useBountyVoting(
  daoId: string,
  proposal: BountyProposal
): {
  handleVote: (vote: VoteAction) => void;
  loading: boolean;
} {
  const router = useRouter();
  const [showModal] = useModal(ConfirmActionModal);
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();

  const [{ loading }, handleVote] = useAsyncFn(
    async (vote: VoteAction) => {
      const res = await showModal({
        title: 'Confirm Your Vote',
      });

      if (res?.length) {
        setVoteActionSource(pathname);
        await nearService?.vote(daoId, proposal.proposalId, vote, res[0]);
        await router.reload();
      }
    },
    [daoId, proposal, router, nearService, pathname, setVoteActionSource]
  );

  return {
    handleVote,
    loading,
  };
}

export function useBountySearch(): {
  handleSearch: (
    val: string
  ) => Promise<PaginationResponse<BountyContext[]> | null>;
  loading: boolean;
} {
  const router = useRouter();
  const daoId = router.query.dao as string;
  const { accountId } = useWalletContext();
  const cancelTokenRef = useRef<CancelTokenSource | null>(null);

  const [{ loading }, handleSearch] = useAsyncFn(async query => {
    if (cancelTokenRef.current) {
      cancelTokenRef.current?.cancel('Cancelled by new req');
    }

    const { CancelToken } = axios;
    const source = CancelToken.source();

    cancelTokenRef.current = source;

    return SputnikHttpService.findBountyContext({
      daoId: daoId || '',
      accountId,
      query,
      cancelToken: source.token,
    });
  }, []);

  return {
    handleSearch,
    loading,
  };
}

export function useHideBounty(): {
  handleSubmit: () => void;
  handleSelect: (id: string) => void;
  selected: string[];
  loading: boolean;
} {
  const router = useRouter();
  const showHidden = router.query?.bountyFilter === 'hidden';
  const daoId = router.query.dao as string;
  const { accountId, nearService, pkAndSignature } = useWalletContext();
  const [selected, setSelected] = useState<string[]>([]);

  const [{ loading }, handleSubmit] = useAsyncFn(async () => {
    if (!pkAndSignature) {
      return;
    }

    try {
      const { publicKey, signature } = pkAndSignature;

      if (publicKey && signature) {
        await SputnikHttpService.toggleBountyContexts({
          accountId,
          publicKey,
          signature,
          daoId,
          ids: selected,
          isArchived: !showHidden,
        });

        await router.replace(router.asPath, undefined, {
          shallow: false,
        });

        setSelected([]);
      }
    } catch (err) {
      showNotification({
        type: NOTIFICATION_TYPES.ERROR,
        lifetime: 20000,
        description: err.message,
      });
    }
  }, [
    selected,
    daoId,
    showHidden,
    accountId,
    router,
    nearService,
    pkAndSignature,
  ]);

  const handleSelect = useCallback(
    (id: string) => {
      if (selected.includes(id)) {
        setSelected(selected.filter(item => item !== id));
      } else {
        setSelected([...selected, id]);
      }
    },
    [selected]
  );

  return {
    handleSubmit,
    handleSelect,
    selected,
    loading,
  };
}
