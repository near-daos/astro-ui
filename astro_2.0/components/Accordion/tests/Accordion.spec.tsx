import { render } from 'jest/testUtils';
import { fireEvent } from '@testing-library/dom';

import { Accordion } from 'astro_2.0/components/Accordion';
import React from 'react';

jest.mock('react', () => {
  return {
    ...jest.requireActual('react'),
    useState: jest.fn(),
  };
});

describe('Accordion', () => {
  it('Should render component', () => {
    const setOpen = jest.fn();

    jest
      .spyOn(React, 'useState')
      .mockImplementationOnce(() => [false, setOpen]);

    const { container } = render(<Accordion title="Hello">World</Accordion>);

    expect(container).toMatchSnapshot();
  });

  it('Should open accordion', () => {
    const setOpen = jest.fn();

    jest
      .spyOn(React, 'useState')
      .mockImplementationOnce(() => [false, setOpen]);

    const { getByTestId } = render(<Accordion title="Hello">World</Accordion>);

    fireEvent.click(getByTestId('accordion-header'));

    expect(setOpen).toBeCalledWith(true);
  });
});
