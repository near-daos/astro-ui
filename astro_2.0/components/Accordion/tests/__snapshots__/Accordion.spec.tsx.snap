// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Should render component 1`] = `
<div>
  <div
    class="root"
  >
    <div
      class="header"
      data-testid="accordion-header"
      role="button"
      tabindex="0"
    >
      Hello
      <button
        class="iconButton medium icon"
        type="button"
      >
        <svg>
          <use
            href="/_next/undefined"
          />
        </svg>
      </button>
    </div>
    <div
      class="contentContainer"
      data-testid="accordion-container"
      style="height: 0px;"
    >
      <div
        class="content"
      >
        World
      </div>
    </div>
  </div>
</div>
`;
